mikee47/Sming-UPnP
# Create control point classes

Wrap all boilerplate stuff in classes, add `Basic_ControlPoint` sample to demonstrate. Applications don't need to touch XML at all.

To do:

* [x] Generate device/service class header from XML document tree. Documents must be organised as on target device.
* [x] Implement `ControlPoint::beginSearch` methods for pre-defined device classes. Does not require description fetch.
* [x] Implement action processing (SOAP stuff)
* [x] Add back description fetching stage so we get correct friendlyName and endpoint URLs.
* [x] Use a custom stream class to process description in chunks so we don't need so much memory.
* [x] Add `allowedValue`s as constants
* [x] Check/fix handling of SOAP errors from remote endpoint
* [x] Add `make` target for scanning local network and pulling in device description trees. Can do this very easily as a host application with recursive fetches for all description files, writing them out to directory tree. There are existing tools to do this but harder to integrate.
* [x] Organise generated directory structure and namespaces to match URN
* [x] Refactor for consistency with hosted devices.
* [x] Implement support for asynchronous action request handling
* [x] Add generator for hosted devices and services. Two files will be generated:
    - [x] The first file will contain class templates for devices and services.  Its purpose is to translate SOAP requests into typed methods.
    - [x] The second file will implement concrete clases using the templates, and must be edited by implementor to respond to requests.

Note: Added python dependency for `lxml`.

The way this is going we follow these steps to implement support for any UPnP device:

1. Perform network scan to pull in device description files into hierarchical directory structure.
2. Update project component.mk file to build header file (s)
3. #include appropriate header files in application
4. Issue search on pre-defined device or service class
5. Handle found device in callback with steps such as:
   - Verifying device/service fields to see if it's the actual device we want (can cancel the search if so)
   - Storing the device object for later use
   - Issuing commands the device (calling action methods on Service object)

Having the class information pre-defined means there is no requirement to fetch or parse those large description files **provided** we are targeting a specific device. This does not cover the general case since we still need the device description to obtain the correct endpoints (and friendlyName plus a few other attributes).

Future work:

* Add event notification support
* Add action argument validation: state variables can have things such as allowedValue, minimum, maximum. This is perhaps something that hosted devices should do more than control devices, since the endpoint will validate.
* Add global exception/error handling callback(s) for framework with appropriate Error codes.
* Consider adding automatic search timeout, cancelling search or sending additional msearch requests. Should we add a search queue?
* Add support for large (streamable) result fields. For example, ContentDirectory results are in DIDL format and can be large. The declared data type is `string` but this would be better handled by a custom stream type to add the additional layer(s) of abstraction. This could be done with additional (custom) attributes in the schema or by using custom data types. The user API would be implemented in `ActionResponse` and `ActionRequest` classes.
